Cassandra-19010 Remove sstable format 'nc' and make 5.0 'nb' and 'oa'

bereng opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

bereng committed:
Default to nb instead of nc for sstable formats

bereng committed:
CIRCLE TO PAID DO NOT MERGE

bereng commented:
Ci [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/1100/workflows/d4356670-cfff-411b-9fde-78aed266862c)
CI [j17](https://app.circleci.com/pipelines/github/bereng/cassandra/1100/workflows/edf84b12-afd8-4ebe-917b-966f554482fe)

bereng commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -458,7 +458,7 @@ protected SSTableReader(Builder<?, ?> builder, Owner owner)
         this.first = builder.getFirst();
         this.last = builder.getLast();
-        this.bounds = AbstractBounds.strictlyWrapsAround(first.getToken(), last.getToken())
+        this.bounds = first == null || last == null || AbstractBounds.strictlyWrapsAround(first.getToken(), last.getToken())
```
@jacek-lewandowski does this make sense to you? Ping me if you need more context

jacek-lewandowski replied:
ok, this is valid +1

bereng replied:
Just for future context: DTest scrub_test.TestScrub.test_standalone_scrub_essential_files_only exercises a path where we scrub sstables where some of the sstable files are missing. This can make those vars null so it's a rare but valid code path.

bereng commented:
Note to self: Remove `test/data/legacy-sstables/nc`

michaelsembwever reviewed (approved):
though i cannot answer [this](https://github.com/apache/cassandra/pull/2880/files#r1386252646)…

jacek-lewandowski commented on test/unit/org/apache/cassandra/io/sstable/metadata/MetadataSerializerTest.java:
```diff
@@ -85,7 +86,9 @@ public void testSerialization() throws IOException
             for (MetadataType type : MetadataType.values())
             {
-                assertEquals(originalMetadata.get(type), deserialized.get(type));
+                if ((type != MetadataType.STATS) || latestVersion.hasImprovedMinMax())
```
is it needed?

bereng replied:
It's the same problem as before where `first` and `last` are null for `nb`. Let's solve that underlying issue first and then we'll know if we still need this.

jacek-lewandowski replied:
got it, it is another consequence of the fact that getLatestVersion does not return the latest version

jacek-lewandowski commented on test/unit/org/apache/cassandra/io/sstable/LegacySSTableTest.java:
```diff
@@ -101,7 +101,7 @@ public class LegacySSTableTest
     {
-        String[] versions = {"oa", "da", "nc", "nb", "na", "me", "md", "mc", "mb", "ma"};
+        String[] versions = {"oa", "da", "nb", "na", "me", "md", "mc", "mb", "ma"};
         return Arrays.stream(versions).filter((v) -> v.compareTo(BigFormat.getInstance().getLatestVersion().toString()) <= 0).toArray(String[]::new);
```
btw. why is this filtering? It seems like it does not filter anything. Moreover if feels incorrect as it assumes `BigFormat`

bereng replied:
`oa` needs to be filtered out unless you do `test-oa`. So in compatibility mode you can go up to nb, out of it you can go up to oa.

jacek-lewandowski replied:
so it is just because we cannot read oa sstables

jacek-lewandowski commented:
+1, the remaining issues are to be discussed in https://issues.apache.org/jira/browse/CASSANDRA-19012

frankgh reviewed (approved):
+1, looks good to me

bereng commented:
Rebased and squashed [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/1105/workflows/e557a641-d861-45af-91ee-f3b451d385ab) and [j17](https://app.circleci.com/pipelines/github/bereng/cassandra/1105/workflows/ed49ce78-e9dd-4b19-bbd4-47ba8255537c). Only an unrelated Circle env problem
